packages/react-reconciler/src/ReactFiberWorkLoop.new.js

```diff
@@ -2645,44 +2676,239 @@ export function enqueuePendingPassiveProfilerEffect(fiber: Fiber): void {
+  const updateQueue: FunctionComponentUpdateQueue | null = (fiber.updateQueue: any);
+  const lastEffect = updateQueue !== null ? updateQueue.lastEffect : null;
+  if (lastEffect !== null) {
+    const firstEffect = lastEffect.next;
+    let effect = firstEffect;
+    do {
+      const {next, tag} = effect;
+
+      if (
+        (tag & HookPassive) !== NoHookEffect &&
+        (tag & HookHasEffect) !== NoHookEffect
+      ) {
+        if (__DEV__) {
+          setCurrentDebugFiberInDEV(fiber);
+          if (
+            enableProfilerTimer &&
+            enableProfilerCommitHooks &&
+            fiber.mode & ProfileMode
+          ) {
+            startPassiveEffectTimer();
+            invokeGuardedCallback(
+              null,
+              invokePassiveEffectCreate,
+              null,
+              effect,
+            );
+            recordPassiveEffectDuration(fiber);
+          } else {
+            invokeGuardedCallback(
+              null,
+              invokePassiveEffectCreate,
+              null,
+              effect,
+            );
+          }
+          if (hasCaughtError()) {
+            invariant(fiber !== null, 'Should be working on an effect.');
+            const error = clearCaughtError();
+            captureCommitPhaseError(fiber, error);
+          }
+          resetCurrentDebugFiberInDEV();
+        } else {
+          try {
+            const create = effect.create;
+            if (
+              enableProfilerTimer &&
+              enableProfilerCommitHooks &&
+              fiber.mode & ProfileMode
+            ) {
+              try {
+                startPassiveEffectTimer();
+                effect.destroy = create();
+              } finally {
+                recordPassiveEffectDuration(fiber);
+              }
+            } else {
+              effect.destroy = create();
+            }
+          } catch (error) {
+            invariant(fiber !== null, 'Should be working on an effect.');
+            captureCommitPhaseError(fiber, error);
+          }
+        }
+      }
+
+      effect = next;
+    } while (effect !== firstEffect);
   }
 }
 
-function invokePassiveEffectCreate(effect: HookEffect): void {
-  const create = effect.create;
-  effect.destroy = create();
+function flushPassiveUnmountEffects(firstChild: Fiber): void {
+  let fiber = firstChild;
+  while (fiber !== null) {
+    const deletions = fiber.deletions;
+    if (deletions !== null) {
+      for (let i = 0; i < deletions.length; i++) {
+        const fiberToDelete = deletions[i];
+        // If this fiber (or anything below it) has passive effects then traverse the subtree.
+        const primaryEffectTag = fiberToDelete.effectTag & PassiveMask;
+        const primarySubtreeTag = fiberToDelete.subtreeTag & PassiveSubtreeTag;
+        if (
+          primarySubtreeTag !== NoSubtreeTag ||
+          primaryEffectTag !== NoEffect
+        ) {
+          flushPassiveUnmountEffects(fiberToDelete);
```

Comment:
Ah I didn't notice this was a recursive call.

I think this should be a different function, analagous to `commitNestedUnmounts`. They do slightly different things. This new function gets passed a fiber from the old tree, not the new tree. So for example, it doesn't have to check for nested deletions, because you're already inside a deleted tree.

(This might be related to the `didBailout` thing?)

Reply:
They don't really do different things. They traverse a tree and execute any passive effects that are scheduled.

> This might be related to the `didBailout` thing?

Which thing?

Reply:
The key difference is that the top-most `flushPassiveUnmountEffects` gets passed a "previous" fiber, from the old tree,  whereas the new function I'm proposing gets passed a "next" fiber, from the just-finished tree.

It goes back to the `didBailout` conversation from the last PR. The `effectTag` and `subtreeTag` on the "previous" tree represents a render that already happened. So we don't want to rely on those being correct for the current render, because that requires us to reset them in the commit phase. (However, it's fine to read the static part `effectTag`, and their corresponding accumulation in `subtreeTag`, since aren't specific to a particular render.)

Same with the `deletions` field. We shouldn't read `previous.deletions` because those deletions aren't part of the current commit phase; we already processed those. In other words, we shouldn't have to null out `finishedWork.deletions` once we're done processing them; I think it's only necessary right now because of the issue I'm describing.

Reply:
> So for example, it doesn't have to check for nested deletions, because you're already inside a deleted tree.

This is a reasonable point. I can split the two functions apart, if you think the added bytes is better than the unnecessary null check.

> In other words, we shouldn't have to null out `finishedWork.deletions` once we're done processing them

Clearing the `deletions` array was unnecessary. I'll remove that line.

Reply:
@acdlite Let me know if ae37c7c addresses your suggestion here?